MIGRATIONS.md

```diff
@@ -0,0 +1,356 @@
+
+While `ember-modifier` is providing this in a backwards-compatible way so that you can migrate on a modifier-by-modifier basis, this change in semantics is *probably* a breaking change for people using your modifiers! If it is (as discussed in detail below), you should release a new major version after updating to use the new semantics. Additionally, for addon maintainers, note that if you release a version of your package after upgrading to 3.2.0, anyone using your modifiers will see deprecations, so you should plan to fix them *before* making that release.
+
+The point is to give you a path to decouple the upgrade and the migration work:
+
+- upgrade to 3.2.0
+- migrate to the new APIs
+- cut a new release (as a breaking change if necessary)
+- upgrade to 4.0.0 when it comes out (no changes required for your code or your consumers)
+
+
+### Function-based modifiers
+
+For function-based modifiers, you simply need to pass a new options argument to the `modifier()` call.
+
+Previously, you would write this:
+
+```js
+export default modifier((el, pos, named) => {
+  // ...
+});
+```
+
+The exactly equivalent behavior with the new options object is:
+
+```js
+export default modifier((el, pos, named) => {
+  // ...
+}, { eager: true });
+```
+
+To migrate to the behavior required for v4, you need to pass `{ eager: false }`:
+
+```js
+export default modifier((el, pos, named) => {
+  // ...
+}, { eager: false });
+```
+
+
+#### When is this a breaking change?
+
+Previously, any time any argument passed to a modifier changed, Ember would re-run the modifier. This was true whether or not you ever used the argument. This means that updating to use the `{ eager: false }` version is a breaking change for your consumers *unless* you unconditionally used all the named and positional arguments arguments to your modifier in previously.[^technically]
+
+
+### Class-based modifiers
+
+For class-based modifiers,  you need to migrate to the new `modify()` API. *All* of the old lifecycle hooks are deprecated, as are the `element`, `args`, and `isDestroying` and `isDestroyed` fields. Additionally, the  new `modify()` hook which replaces the previous lifecycle hooks is *lazy*, like auto-tracking in general. After installation, it will only be re-executed when some tracked state it uses actually changes, where previously it.
+
+
+#### Lifecycle hooks
+
+##### `didInstall()`
+
+For one-time setup done with the element in `didInstall()`, do the same setup in `modify()`. If it is cheap, you can just let it happen each time. If it is expensive, or if the operation is not [idempotent](https://en.wikipedia.org/wiki/Idempotence), you can set a flag to avoid doing it again:
```

Comment:
Might be nice to have an example in the normal docs (not these migration docs though probably) of how you can do this same kinda "flag" deal but still have it be dependent on some incoming state by using tracked cache (e.g. `import { createStorage, getValue, setValue } from 'ember-tracked-storage-polyfill'`).